packages/realm/src/schema/types.ts

```diff
@@ -71,30 +101,221 @@ export type CanonicalObjectSchemaProperty = {
+ * type is provided as guidance. (Exact errors will always be shown when creating a
+ * {@link Realm} instance if the schema is invalid.)
+ */
+export type PropertySchema = {
+  /**
+   * The type of the property.
+   */
+  type: PropertyTypeName;
+  /**
+   * The type of the elements in the collection if `type` is a {@link CollectionPropertyTypeName},
+   * or the specific Realm object type if `type` is a {@link RelationshipPropertyTypeName}.
+   */
+  objectType?: PrimitivePropertyTypeName | UserTypeName;
+  /**
+   * The name of the property of the object specified in `objectType` that creates this link.
+   */
+  property?: string;
+  /**
+   * Whether to allow `null` or `undefined` to be assigned to the property; or in the
+   * case of a collection, to be assigned to its elements. (Realm object types in lists
+   * and sets cannot be optional.)
+   *
+   * Default value: See the requirements in the documentation for this type.
+   */
+  optional?: boolean;
+  /**
+   * Whether the property should be indexed. (Only supported if `type` is `"string"`,
+   * `"int"`, or `"bool"`).
+   *
+   * Default value: `false` if the property is not a primary key, otherwise `true`.
+   */
+  indexed?: boolean;
+  /**
+   * The name to be persisted in the Realm file if it differs from the already-defined
+   * JavaScript/TypeScript (JS/TS) property name. This is useful for allowing different
+   * naming conventions than what is persisted in the Realm file. Reading and writing
+   * properties must be done using the JS/TS name, but queries can use either the JS/TS
+   * name or the persisted name.
+   */
+  mapTo?: string;
+  /**
+   * The default value that the property will be set to when created.
+   */
+  default?: unknown;
+};
+
+/**
+ * Keys used in the property schema that are common among all variations of {@link PropertySchemaStrict}.
+ */
+type PropertySchemaCommon = {
+  indexed?: boolean;
+  mapTo?: string;
+  default?: unknown;
+};
+
+/**
+ * The strict schema for specifying the type of a specific Realm object property.
+ *
+ * Unlike the less strict {@link PropertySchema}, this type precisely defines the type
+ * requirements and their allowed combinations; however, TypeScript error messages tend
+ * to be more misleading. {@link PropertySchema} is recommended for that reason, but the
+ * strict type is provided as guidance.
+ *
+ * @see {@link PropertySchema} for a textual explanation of the requirements defined here,
+ *   as well as documentation for each property.
+ */
+export type PropertySchemaStrict = PropertySchemaCommon &
+  (
+    | {
+        type: Exclude<PrimitivePropertyTypeName, "mixed">;
+        optional?: boolean;
+        // TODO: Do we want TS errors for PropertySchemaStrict when `objectType` and
+        //       `property` exist on the schema as `undefined`? (The parser allows this)
+        // objectType?: undefined;
+        // property?: undefined;
```

Comment:
I think so, if it's doable.  We should leverage TypeScript to guide our users to creating the correct schema if possible.  Better to get a type error than a runtime error.

Reply:
I'm not sure we should treat properties explicitly set to undefined differently from missing properties. I know ts does sometimes, but that seems surprising to me. It makes it harder to do some things that Just Work in js because `obj.blah` behaves the same regardless of whether blah is missing or set to undefined.

Reply:
Users would most likely not explicitly define a `myKey: undefined`, but was thinking more if this strict type should align with the rules of the parser (allowing `undefined`) or just not allow the field at all. Mathias, so are you saying you would vote for keeping the fields?

Reply:
Since this is primarily for documentation rather than the type users will actually use, I think we should leave them out here. Putting optional undefined fields in seems likely to cause more confusion than it would solve. Anyone who needs to pass undefined fields will probably just assume it works and try it. Especially since the type we use in APIs accepts them.